datafusion/core/src/physical_plan/hash_join.rs

```diff
@@ -947,6 +951,27 @@ macro_rules! equal_rows_elem {
     }};
 }
 
+macro_rules! equal_rows_elem_with_string_dict {
+    ($key_array_type:ident, $l: ident, $r: ident, $left: ident, $right: ident, $null_equals_null: ident) => {{
```

Comment:
For dictionaries, I think `$left` and `$right` are actually indexes into the *keys* array, and then the keys array contains the corresponding index into `values`.

```text
┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─                                                
  ┌─────────────────┐  ┌─────────┐ │     ┌─────────────────┐                       
│ │        A        │  │    0    │       │        A        │     values[keys[0]]   
  ├─────────────────┤  ├─────────┤ │     ├─────────────────┤                       
│ │        D        │  │    2    │       │        B        │     values[keys[1]]   
  ├─────────────────┤  ├─────────┤ │     ├─────────────────┤                       
│ │        B        │  │    2    │       │        B        │     values[keys[2]]   
  ├─────────────────┤  ├─────────┤ │     ├─────────────────┤                       
│ │        C        │  │    1    │       │        D        │     values[keys[3]]   
  ├─────────────────┤  └─────────┘ │     └─────────────────┘                       
│ │        E        │     keys                                                     
  └─────────────────┘              │                                               
│       values                             Logical array                           
 ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘          Contents                             
                                                                                   
          DictionaryArray                                                          
             length = 4                                                            
                                                                                   
```

In other words, I think you need to compare the values using something like:
https://github.com/AssHero/arrow-datafusion/blob/hashjoin/datafusion/common/src/scalar.rs#L338-L361